File: cordapp-cpk2/src/main/java/net/corda/plugins/cpk2/CordappExtension.java
package net.corda.plugins.cpk2;

import org.gradle.api.Action;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.Property;
import org.gradle.api.provider.ProviderFactory;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.Nested;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.inject.Inject;

import static net.corda.plugins.cpk2.CordappUtils.CORDAPP_SEALING_SYSTEM_PROPERTY_NAME;
import static net.corda.plugins.cpk2.CordappUtils.PLATFORM_VERSION_X;

/**
 * Top-level CorDapp attributes.
 */
public class CordappExtension {
    private final Property<Integer> targetPlatformVersion;
    private final Property<Integer> minimumPlatformVersion;
    private final CordappData contract;
    private final CordappData workflow;
    private final Signing signing;
    private final Property<Boolean> sealing;
    private final Property<String> bndVersion;
    private final Property<String> osgiVersion;
    private final Property<String> jetbrainsAnnotationsVersion;
    private final Property<String> hashAlgorithm;

    @Inject
    public CordappExtension(
        @NotNull ObjectFactory objects,
        @NotNull ProviderFactory providers,
        @Nullable String osgiVersion,
        @Nullable String bndVersion,
        @Nullable String jetbrainsAnnotationsVersion
    ) {
        targetPlatformVersion = objects.property(Integer.class);
        minimumPlatformVersion = objects.property(Integer.class).convention(PLATFORM_VERSION_X);
        contract = objects.newInstance(CordappData.class);
        workflow = objects.newInstance(CordappData.class);
        signing = objects.newInstance(Signing.class);
        sealing = objects.property(Boolean.class).convention(
            providers.systemProperty(CORDAPP_SEALING_SYSTEM_PROPERTY_NAME).orElse("true").map(Boolean::parseBoolean)
        );
        this.bndVersion = objects.property(String.class).convention(bndVersion);
        this.osgiVersion = objects.property(String.class).convention(osgiVersion);
        this.jetbrainsAnnotationsVersion = objects.property(String.class).convention(jetbrainsAnnotationsVersion);
        hashAlgorithm = objects.property(String.class).convention("SHA-256");
    }

    @Input
    @NotNull
    public Property<Integer> getTargetPlatformVersion() {
        return targetPlatformVersion;
    }

    @Input
    @NotNull
    public Property<Integer> getMinimumPlatformVersion() {
        return minimumPlatformVersion;
    }

    /**
     * @return CorDapp Contract distribution information.
     */
    @Nested
    @NotNull
    public CordappData getContract() {
        return contract;
    }

    /**
     * @return CorDapp Workflow (flows and services) distribution information.
     */
    @Nested
    @NotNull
    public CordappData getWorkflow() {
        return workflow;
    }

    /**
     * @return Optional parameters for ANT signJar tasks to sign CorDapps.
     */
    @Nested
    @NotNull
    public Signing getSigning() {
        return signing;
    }

    /**
     * @return Optional marker to seal all packages in the JAR.
     */
    @Input
    @NotNull
    public Property<Boolean> getSealing() {
        return sealing;
    }

    @Input
    @NotNull
    public Property<String> getBndVersion() {
        return bndVersion;
    }

    @Input
    @NotNull
    public Property<String> getOsgiVersion() {
        return osgiVersion;
    }

    @Input
    @NotNull
    public Property<String> getJetbrainsAnnotationsVersion() {
        return jetbrainsAnnotationsVersion;
    }

    /**
     * @return This property only provides the default value for {@link CPKDependenciesTask#getHashAlgorithm},
     * which is annotated with @Input. Hence this is not a task input itself.
     */
    @Input
    @NotNull
    public Property<String> getHashAlgorithm() {
        return hashAlgorithm;
    }

    public void contract(@NotNull Action<? super CordappData> action) {
        action.execute(contract);
    }

    public void workflow(@NotNull Action<? super CordappData> action) {
        action.execute(workflow);
    }

    public void signing(@NotNull Action<? super Signing> action) {
        action.execute(signing);
    }

    public void targetPlatformVersion(@Nullable Integer value) {
        targetPlatformVersion.set(value);
    }

    public void minimumPlatformVersion(@Nullable Integer value) {
        minimumPlatformVersion.set(value);
    }
}

File: cordapp-cpk2/src/main/kotlin/net/corda/plugins/cpk2/CordappExtension.kt
@file:JvmName("CordappProperties")
package net.corda.plugins.cpk2

import org.gradle.api.Action
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
import org.gradle.api.provider.ProviderFactory
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.Nested
import org.gradle.api.tasks.TaskInputs
import javax.inject.Inject

/**
 * Registers these [CordappExtension] properties as task inputs,
 * because Gradle cannot "see" their `@Input` annotations yet.
 */
fun TaskInputs.nested(nestName: String, cordapp: CordappExtension) {
    property("${nestName}.targetPlatformVersion", cordapp.targetPlatformVersion)
    property("${nestName}.minimumPlatformVersion", cordapp.minimumPlatformVersion)
    nested("${nestName}.contract", cordapp.contract)
    nested("${nestName}.workflow", cordapp.workflow)
    nested("${nestName}.signing", cordapp.signing)
    property("${nestName}.sealing", cordapp.sealing)
    property("${nestName}.bndVersion", cordapp.bndVersion)
}

@Suppress("UnstableApiUsage", "Unused", "PlatformExtensionReceiverOfInline")
open class CordappExtension @Inject constructor(
    objects: ObjectFactory,
    providers: ProviderFactory,
    osgiVersion: String,
    bndVersion: String
) {
    /**
     * Top-level CorDapp attributes
     */
    @get:Input
    val targetPlatformVersion: Property<Int> = objects.property(Int::class.java)

    @get:Input
    val minimumPlatformVersion: Property<Int> = objects.property(Int::class.java)
            .convention(PLATFORM_VERSION_X)

    /**
     * CorDapp Contract distribution information.
     */
    @get:Nested
    val contract: CordappData = objects.newInstance(CordappData::class.java)

    /**
     * CorDapp Workflow (flows and services) distribution information.
     */
    @get:Nested
    val workflow: CordappData = objects.newInstance(CordappData::class.java)

    /**
     * Optional parameters for ANT signJar tasks to sign CorDapps.
     */
    @get:Nested
    val signing: Signing = objects.newInstance(Signing::class.java)

    /**
     * Optional marker to seal all packages in the JAR.
     */
    @get:Input
    val sealing: Property<Boolean> = objects.property(Boolean::class.java).convention(
        providers.systemProperty(CORDAPP_SEALING_SYSTEM_PROPERTY_NAME).orElse("true").map(String::toBoolean)
    )

    @get:Input
    val bndVersion: Property<String> = objects.property(String::class.java).convention(bndVersion)

    @get:Input
    val osgiVersion: Property<String> = objects.property(String::class.java).convention(osgiVersion)

    /**
     * This property only provides the default value for [CPKDependenciesTask.hashAlgorithm], which is
     * annotated with @Input. Hence this is not a task input itself.
     */
    val hashAlgorithm: Property<String> = objects.property(String::class.java).convention("SHA-256")

    fun contract(action: Action<in CordappData>) {
        action.execute(contract)
    }

    fun workflow(action: Action<in CordappData>) {
        action.execute(workflow)
    }

    fun signing(action: Action<in Signing>) {
        action.execute(signing)
    }

    fun targetPlatformVersion(value: Int?) {
        targetPlatformVersion.set(value)
    }

    fun minimumPlatformVersion(value: Int?) {
        minimumPlatformVersion.set(value)
    }
}
